apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1.IntegrationTests/Emulator/metadata_server.py

```diff
@@ -0,0 +1,353 @@
+    text = ''
+    if isinstance(o, list):
+      for val in o:
+        text += convertToTextRecursive(prefix, val)
+    elif isinstance(o, IndexedList):
+      for i in range(len(o)):
+        text += convertToTextRecursive(prefix + ('/' if len(prefix) != 0 else '') + unicode(i),
+                                       o[i])
+    elif isinstance(o, OrderedDict):
+      for key in o:
+        if excludeKeyInRecursiveGet(key):
+          continue
+        text += convertToTextRecursive(prefix + ('/' if len(prefix) != 0 else '') + key,
+                                       o[key])
+    elif prefix:
+      text += prefix + ' ' + unicode(o) + '\n'
+    else:
+      text += unicode(o)
+
+    return text
+
+  alt = request.args.get('alt', '')
+  recursive = asBool(request.args.get('recursive', ''))
+  timeout_sec = asInt(request.args.get('timeout_sec', ''), -1)
+  wait_for_change = asBool(request.args.get('wait_for_change', ''))
+
+  components = path.split('/')
+  result = all_data
+  ends_with_slash = False
+  is_token_request = False
+
+  for component in components:
+    if len(component) == 0:
+      ends_with_slash = True
+      break
+
+    is_token_request = (component == 'token')
+    ends_with_slash = False
+    
+    if isinstance(result, IndexedList):
+      try:
+        index = int(component)
+      except ValueError:
+        return not_found_error
+      
+      if 0 <= index and index < len(result):
+        result = result[index]
+      else:
+        return not_found_error
+    else:
+      component = component.replace(unicode(numeric_project_id), numeric_project_id_token)
+      component = component.replace(project_id, project_id_token)
+      if not component in result:
+        return not_found_error
+
+      result = result[component]
+
+  if result == numeric_project_id_token:
+    result = numeric_project_id
+
+  if is_token_request:
+    # Even though this acts like a directory, trailing slashes are not allowed for tokens.
+    if ends_with_slash:
+      return not_found_error
+
+    is_directory = True
+    is_json = (alt != 'text')
+    recursive = True
+  else:
+    # recursive directory requests and instance/tags default to json
+    is_directory = isinstance(result, OrderedDict) or isinstance(result, IndexedList) or isinstance(result, list)
+    is_json = (alt == 'json' or ((component == 'tags' or (recursive and is_directory)) and alt != 'text'))
+    if ends_with_slash and not is_directory:
+      return not_found_error
+
+  if wait_for_change:
+    # When waiting for changes on non-list directories, it must be a recursive request
+    if is_directory and not isinstance(result, list) and not recursive:
+      return bad_request_error
+
+    if timeout_sec < 0:
+      return formatResponse('Error: indefinitely waiting for changes is not supported',
+                            400, 
+                            common.content_type_html)
+    time.sleep(timeout_sec)
+
+  if recursive:
```

Comment:
this method is pretty big in general, it feels like it could benefit from breaking pieces up into methods a bit where possible